contrib/check-semver.sh

```diff
@@ -0,0 +1,202 @@
+    local crate="$1"
+    local variant="$2"
+
+    echo "Running cargo semver-checks for $crate $variant"
+    # Hack to not fail on errors.
+    # This is necessary since cargo semver-checks will fail if the
+    # semver check fails.
+    # We check that manually later.
+    set +e
+    cargo +"$NIGHTLY" semver-checks -v --baseline-rustdoc "$crate-master-$variant.json" --current-rustdoc "$crate-current-$variant.json" > "$crate-$variant-semver.txt" 2>&1
+    set -e
+}
+
+# The following function uses cargo doc to generate JSON files that
+# cargo semver-checks can use.
+# - no-default-features: generate JSON doc files with no default features.
+generate_json_files_no_default_features() {
+    local crate="$1"
+    local version="$2"
+
+    echo "Running cargo doc no-default-features for $crate $version"
+    run_cargo_doc --no-default-features -p "$crate"
+
+    # replace _ for - in crate name.
+    # This is necessary since some crates have - in their name
+    # which will be converted to _ in the output file by cargo doc.
+    mv "target/doc/${crate//-/_}.json" "$crate-$version-no-default-features.json"
+}
+# - all-features: generate JSON doc files with all features.
+generate_json_files_all_features() {
+    local crate="$1"
+    local version="$2"
+
+    echo "Running cargo doc all-features for $crate $version"
+    run_cargo_doc --all-features -p "$crate"
+
+    # replace _ for - in crate name.
+    # This is necessary since some crates have - in their name
+    # which will be converted to _ in the output file by cargo doc.
+    mv -v "target/doc/${crate//-/_}.json" "$crate-$version-all-features.json"
+}
+# - alloc: generate JSON doc files with the alloc feature.
+generate_json_files_features_alloc() {
+    local crate="$1"
+    local version="$2"
+
+    echo "Running cargo doc --features alloc for $crate $version"
+    run_cargo_doc --no-default-features --features alloc -p "$crate"
+
+    # replace _ for - in crate name.
+    # This is necessary since some crates have - in their name
+    # which will be converted to _ in the output file by cargo doc.
+    mv -v "target/doc/${crate//-/_}.json" "$crate-$version-alloc.json"
+}
+
+# Check if there are breaking changes.
+# We loop through all the generated files and check if there is a FAIL
+# in the cargo semver-checks output.
+# If we detect a fail, we create an empty file semver-break.
+# If the following CI step finds this file, it will add:
+# 1. a comment on the PR.
+# 2. a label to the PR.
+check_for_breaking_changes() {
+    for file in *semver.txt; do
+        echo "Checking $file"
+        if grep -q "FAIL" "$file"; then
+            echo "You have introduced changes to the public API"
+            echo "FAIL found in $file"
+            # flag it as a breaking change
+            # Handle the case where FAIL is found
+            touch semver-break
+        fi
+    done
+    if ! [ -f semver-break ]; then
+       echo "No breaking changes found"
+    fi
+}
```

Comment:
Now I really don't understand why it's done here and not after the check.

Reply:
This is if all cargo semver-checks got a `FAIL`.

Reply:
Oh, doing it in a loop would create the comment multiple times, right?